source/grid/reference_cell.cc

```diff
@@ -807,6 +807,244 @@ ReferenceCell::gmsh_element_type() const
 
 
 
+namespace
+{
+  // Compute the nearest point to @p on the line segment and the square of its
+  // distance to @p.
+  template <int dim>
+  std::pair<Point<dim>, double>
+  project_to_line(const Point<dim> &x0,
+                  const Point<dim> &x1,
+                  const Point<dim> &p)
+  {
+    Assert(x0 != x1, ExcInternalError());
+    // t is the convex combination coefficient (x = (1 - t) * x0 + t * x1)
+    // defining the position of the closest point on the line (not line segment)
+    // to p passing through x0 and x1. This formula is equivalent to the
+    // standard 'project a vector onto another vector', where each vector is
+    // shifted to start at x0.
+    const double t = ((x1 - x0) * (p - x0)) / ((x1 - x0).norm_square());
+
+    // Only consider points between x0 and x1
+    if (t <= 0)
+      return std::make_pair(x0, x0.distance_square(p));
+    else if (t <= 1)
+      {
+        const auto p2 = x0 + t * (x1 - x0);
+        return std::make_pair(p2, p2.distance_square(p));
+      }
+    else
+      return std::make_pair(x1, x1.distance_square(p));
+  }
+
+  // template base-case
+  template <int dim>
+  std::pair<Point<dim>, double>
+  project_to_quad(const std::array<Point<dim>, 3> & /*vertices*/,
+                  const Point<dim> & /*p*/,
+                  const ReferenceCell /*reference_cell*/)
+  {
+    Assert(false, ExcInternalError());
+    return std::make_pair(Point<dim>(),
+                          std::numeric_limits<double>::signaling_NaN());
+  }
+
+  /**
+   * Compute the nearest point on a quad (in the deal.II sense: i.e., something
+   * with structdim = 2 and spacedim = 3) and the square of that point's
+   * distance to @p.
+   *
+   * If the given point cannot be projected via a normal vector then this
+   * function returns the origin and the largest double precision number.
+   * distance_to_line_square() is in charge of computing the distance to lines.
```

Comment:
Good question (I added more documentation). This function only projects points along the normal vector of the quad (since that's the shortest distance). However, it might turn out that there is no point on the quad which is on the line parallel to the normal vector which passes through `p`: e.g., for a hexahedron, `p = (2, 2, 2)` cannot be projected to any face via any face normal vector.